### PR TITLE
[macOS] Drop SDL windowing implementation

### PR DESCRIPTION
**Pull request by enen92**
## Description
This PR is the result of a lot of small improvements on the native windowing implementation for MacOS as an effort attain feature parity with the SDL counterpart and eventually set it as the default windowing system for Kodi in macOS (while removing the former). Honestly by now, the nativewindowing 
version is far superior to SDL: It supports retina displays (HiDPI), has improved window move and resize operations, app menu, and just feels more natural for regular macOS users. Hopefully it will make it easier to add new features to the platform while ditching one unmaintained dependency.

When XBMC started to be ported to Linux ages ago, SDL was chosen as the input and windowing toolkit to bring the application to the platform. Other ports, like OSX, also "inherited" the implementation. With time and with an effort to split/segregate platform code, SDL started to be removed in favour of specific platform implementations. At this point, only macOS still uses SDL.
The work on native windowing has started a few years ago in a feature branch (https://github.com/xbmc/xbmc/tree/osxremovesdl) by @Memphiz , @Rechi  and others. With the release of recent macs with apple silicon and the troubles to compile Kodi for arm64, @fuzzard has made further improvements to the system and merged it into master as an option (used by default for arm).

This now got to a point where it'd be nice to have users (and team) feedback to consider the change for Alpha3 :)

**List of changes:**

- ~Haven't tested intel x86 (although it should work)~ (works)
- ~Mouse is a bit broken (not positioned on the right place all the time, sometimes is occluded)~ (partial fix https://github.com/xbmc/xbmc/pull/23019, https://github.com/xbmc/xbmc/pull/23036, https://github.com/xbmc/xbmc/pull/23050, https://github.com/xbmc/xbmc/pull/23062 and https://github.com/xbmc/xbmc/pull/23084 and https://github.com/xbmc/xbmc/pull/23194) 
- ~Tilling windows to the left/right is completely broken~ (disabled for now in https://github.com/xbmc/xbmc/pull/23035)
- ~Blanking other displays is broken~ (fixed by https://github.com/xbmc/xbmc/pull/23029 https://github.com/xbmc/xbmc/pull/23262)
- ~Menu items are broken (float on top broken, wrong shortcuts, missing fullscreen toggle callback)~ (fixed by https://github.com/xbmc/xbmc/pull/23031 and https://github.com/xbmc/xbmc/pull/23034 https://github.com/xbmc/xbmc/pull/23192)
- ~Refresh rate switch is untested~ (works)
- ~Window controller could be improved (we're delegating some functionality to Kodi that could probably be done by the controller - @stevehartwell maybe you can contribute this one since you had it sort of working?)~ (https://github.com/xbmc/xbmc/pull/23118)
- Input (e.g. ~copy-paste~ (works), controllers, etc is untested)
- ~Retina resolutions are not properly supported~ (feature in https://github.com/xbmc/xbmc/pull/23259)
- ~NotifyAppFocusChange is not called~ (nuked by https://github.com/xbmc/xbmc/pull/23043)
- ~OcclusionState is unimplemented~ (not needed, nuked in https://github.com/xbmc/xbmc/pull/23039)
- ~Not all kodi shortcuts work...~ (fixed by https://github.com/xbmc/xbmc/pull/23081)
- ~Live resize only refreshes the view on liveresize end~ (fix https://github.com/xbmc/xbmc/pull/23085)
- ~Setfullscreen sometimes do not take insects into consideration~ (https://github.com/xbmc/xbmc/pull/23125)
- ~Blanking windows may leave a black desktop once the main desktop is lost~ (fixed by https://github.com/xbmc/xbmc/pull/23263)
- ~Still some crashes on live resize (has to do with frame being null as we render out of drawrect)~(https://github.com/xbmc/xbmc/pull/23155)
- -MacOS API deprecations~ (https://github.com/xbmc/xbmc/pull/22229)
- ~Fullscreen inconsistencies depending on the code path to toggle fullscreen. Transient resize events with fullscreen dimensions while resizing the window~ (Fix https://github.com/xbmc/xbmc/pull/23187 https://github.com/xbmc/xbmc/pull/22829 https://github.com/xbmc/xbmc/pull/22232 https://github.com/xbmc/xbmc/pull/22217 https://github.com/xbmc/xbmc/pull/22227)
- ~Programmatic move of the Kodi window~ (https://github.com/xbmc/xbmc/pull/22987)
- ~Improve move window between screens~ (https://github.com/xbmc/xbmc/pull/22746)
- ~Improve UI of Kodi window~ (https://github.com/xbmc/xbmc/pull/23107) 
- ~Notch/EdgeInsets support~ (https://github.com/xbmc/xbmc/pull/23125  https://github.com/xbmc/xbmc/pull/22292 https://github.com/xbmc/xbmc/pull/23637)
- ~When the main screen is lost and Kodi moves to the default screen the window is not resized~(https://github.com/xbmc/xbmc/pull/23335)
- ~Restore fullscreen move for the window~ (https://github.com/xbmc/xbmc/pull/23335)
- ~Allow keypresses if mouse is outside of window bounds but the app has focus~ (https://github.com/xbmc/xbmc/pull/23630)

**Fixed Regressions:**
Dialog confirmation not showing after changing resolution : https://github.com/xbmc/xbmc/pull/23432
Wrong resolution labels on android: https://github.com/xbmc/xbmc/pull/23371 https://github.com/xbmc/xbmc/pull/23410 
Low fps on GUI rendering code: https://github.com/xbmc/xbmc/pull/23578
Leftover on resolution change: https://github.com/xbmc/xbmc/pull/23579
Mouse not available when becoming/resigning key: https://github.com/xbmc/xbmc/pull/23659
Enter does not work: https://github.com/xbmc/xbmc/pull/23860

**Testbuilds:**
**MacOS x86:** https://mirrors.kodi.tv/test-builds/osx/x86_64/kodi-20230623-a7068fe7-drop_sdl-30430-x86_64.dmg
**MacOS arm64:** https://mirrors.kodi.tv/test-builds/osx/arm64/kodi-20230623-a7068fe7-drop_sdl-730-arm64.dmg

Thanks to everyone who has worked on the feature for all these years!

🎉 🚀

### CODE REVIEW

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
Jenkins build this please

**Comment by enen92** [comment]
Jenkins build this please

File: xbmc/platform/darwin/osx/CocoaInterface.mm [hunk 1]
```diff
@@ -36,12 +32,7 @@
 
 NSOpenGLContext* Cocoa_GL_GetCurrentContext(void)
 {
-#if defined(HAS_SDL)
-  CWinSystemOSX *winSystem = dynamic_cast<CWinSystemOSX*>(CServiceBroker::GetWinSystem());
-  return winSystem->GetNSOpenGLContext();
-#else
   return [NSOpenGLContext currentContext];
-#endif
 }
```
Comment (fuzzard) [inline xbmc/platform/darwin/osx/CocoaInterface.mm:37]: While you are in here, the following can be removed (header and implementation)

https://github.com/xbmc/xbmc/blob/d6b9189f5ca69886c4195e7695632197e27dd9f6/xbmc/platform/darwin/osx/CocoaInterface.h#L33-L36
https://github.com/xbmc/xbmc/blob/d6b9189f5ca69886c4195e7695632197e27dd9f6/xbmc/platform/darwin/osx/CocoaInterface.mm#L232-L240

Only current use is gated by HAS_SDL which you remove from xbmc/Application.cpp

Comment (enen92) [reply xbmc/platform/darwin/osx/CocoaInterface.mm:37]: thanks for noticing, I plan to review those unused methods. Did you take a deep look or just noticed this one?
I believe a bunch of others might also be unused if SDL is dropped.

Comment (fuzzard) [reply xbmc/platform/darwin/osx/CocoaInterface.mm:37]: Just noticed while dealing with some warnings for SDK bump

**Comment by pkscout** [comment]
I played with the ARM64 build on my M1 MacBook Air for about 20 minutes and didn't run into any issues.  The only thing I couldn't test was moving the windows between screens (my setup with the extra screen isn't available right now).

I assume the ARM64 build is the Apple Silicon native build?

**Comment by enen92** [comment]
Thanks for the feedback @pkscout. Yes you're right, arm64 is the version for apple silicon :)
I think you can also run the x86 version with rosetta

**Comment by enen92** [comment]
Jenkins build this please

**Comment by enen92** [comment]
jenkins build this please

**Review by garbear — APPROVED** [review]
That which is ready to fall, shall ye also push!

**Comment by kambala-decapitator** [comment]
finally tested "in production" on Intel. In general: all works, great job!

minor inconveniences spotted:
- switching to another screen didn't work seamlessly, had to make one more click. Started on a single screen in fullscreen, connected another screen (projector), moved Kodi to the latter via settings -> app switched to window mode and stayed in the first screen, the setting jumped back to Default screen; after selecting another screen once more window moved properly to fullscreen on the other screen.
- starting app takes a bit longer for UI to appear
- cursor size doesn't scale as before (probably because it's in hardware mode)

**Review by kambala-decapitator — APPROVED** [review]

**Comment by enen92** [comment]
> * switching to another screen didn't work seamlessly, had to make one more click. Started on a single screen in fullscreen, connected another screen (projector), moved Kodi to the latter via settings -> app switched to window mode and stayed in the first screen, the setting jumped back to Default screen; after selecting another screen once more window moved properly to fullscreen on the other screen.

I'll try to check, I am not sure the monitor list gets properly updated if you connect hot plug it after startup. Does it work if you start Kodi with the two displays already connected?

> * starting app takes a bit longer for UI to appear

Never noticed this, quite surprising. Afaik rendering code is the same for both windowing systems. Might it be because we render out of the main thread?

> * cursor size doesn't scale as before (probably because it's in hardware mode)

You mean the Kodi or the os cursor? The OS cursor should not be shown when moving the cursor inside the window. Maybe this is because of the retina display support? What is the resolution you have selected? 

> * cursor no longer disappears while movie is playing, might irritate someone

This could be a bug. Again, Kodi or OS cursor?

**Comment by kambala-decapitator** [comment]
> I'll try to check, I am not sure the monitor list gets properly updated if you connect (hot plug) a monitor after startup. Does it work if you start Kodi with the two displays already connected?

will re-check later, but iirc I started Kodi when both screens were connected

> Never noticed this, quite surprising. Afaik rendering code is the same for both windowing systems. Might it be because we render out of the main thread?

now with a single screen connected I don't observe this. Not sure if it was just bad luck or somehow connected to having multiple displays.

> You mean the Kodi or the os cursor? The OS cursor should not be shown when moving the cursor inside the window. Maybe this is because of the retina display support? What is the resolution you have selected?

OS cursor, noticed after moving Kodi to the second screen. Both in menu and during video play. With a single screen I see Kodi cursor.

I don't have retina displays, main screen resolution is 1440p and projector's is 1080p.

> This could be a bug. Again, Kodi or OS cursor?

OS cursor.

### New observations

so last time Kodi was configured to run on the projector, now I started it on a single screen and all was fine, in settings it still says that the selected screen is the projector and resolution is 1440p (looks right). I switched selected display to Default and ended up with this:

Screenshot 2023-07-09 at 06 21 45

Didn't press No, the dialog disappeared and it stayed like that. Then I switched setting from Windowed to Fullscreen and got this:

Screenshot 2023-07-09 at 06 22 26

Restarting Kodi brought it back to normal though.

Also noticed that animations are kinda laggy now, definitely not 60 FPS. Movies run smoothly though.

**Comment by enen92** [comment]
I think both issues (wrong size) and os cursor being shown might have the same cause. Usually we trigger the resize on move callbacks but for some reason some of the callbacks are not being triggered for you (and that also affects the update of the tracking areas). Will try to have a look but I am not sure when I'll have time to dedicate to this again.

Regarding animations or rendering performance unfortunately I don't have any ideas this is something I have not changed (nor understand much to be honest)

**Comment by kambala-decapitator** [comment]
discovered that Kodi doesn't capture keyboard media keys.

- had music player on (JellyFin), it responds to media keys
- stop music, start Kodi, play a movie
- media keys still control JellyFin, not Kodi

**Comment by kambala-decapitator** [comment]
one more issue: changing playback speed via Alt+left/right doesn't work (I did enable sync playback to display). Not sure if it's just a hotkey issue or windowing / Kodi issue.

**Comment by enen92** [comment]
> discovered that Kodi doesn't capture keyboard media keys.
> 
> * had music player on (JellyFin), it responds to media keys
> * stop music, start Kodi, play a movie
> * media keys still control JellyFin, not Kodi
> 
> adding Kodi to Settings - Privacy - Input monitoring didn't help

This more a feature request than an actual bug, I don't think mediakeys used to work (or at least used some legacy API not compatible anymore with modern macs). I tried to implement this while developing this but got to the conclusion it's not straight forward. There are some libraries that implement global system key taps but that even require the application to be whitelisted (e.g. https://github.com/jenghis/mediakeys). Adding new dependencies just because of mediakeys seem a bit counter-produtive as we are trying to get rid of one (SDL) in the first place.
VLC for example implements https://github.com/nevyn/SPMediaKeyTap (ref: https://github.com/videolan/vlc/blob/f7bb59d9f51cc10b25ff86d34a3eff744e60c46e/modules/gui/macosx/imported/SPMediaKeyTap/SPMediaKeyTap.m) but as you can see in the implementation it's a bit "complex"...global handles and even listening to app other app focus events to check when to react or ignore key presses...
This is something that definitely can be implemented as a follow up but I see it more as a later improvement than as a fundamental requirement to drop SDL

**Comment by enen92** [comment]
> one more issue: changing playback speed via Alt+left/right doesn't work (I did enable sync playback to display). Not sure if it's just a hotkey issue or windowing / Kodi issue.

This one might be some issue on our end, I will investigate. Thanks!

**Comment by kambala-decapitator** [comment]
media keys: fully clear, thanks a lot. But users may view it as a regression...

Still, average user would expect a media player app to handle media keys, that's kinda standard thing. Some solution would definitely be needed for 21.x.

**Comment by enen92** [comment]
> > one more issue: changing playback speed via Alt+left/right doesn't work (I did enable sync playback to display). Not sure if it's just a hotkey issue or windowing / Kodi issue.
> 
> This one might be some issue on our end, I will investigate. Thanks!

@kambala-decapitator I cannot reproduce this. After enabling "sync playback to display" the alt+right/left trigger tempo up and down correctly. I am using the laptop keyboard where the option keys (left and right) map to ALT. Are you using some external keyboard or something? Are the keys mapped correctly?

**Comment by kambala-decapitator** [comment]
> > > one more issue: changing playback speed via Alt+left/right doesn't work (I did enable sync playback to display). Not sure if it's just a hotkey issue or windowing / Kodi issue.
> > 
> > 
> > This one might be some issue on our end, I will investigate. Thanks!
> 
> @kambala-decapitator I cannot reproduce this. After enabling "sync playback to display" the alt+right/left trigger tempo up and down correctly. I am using the laptop keyboard where the option keys (left and right) map to ALT. Are you using some external keyboard or something? Are the keys of this hypotethical external keyboard mapped correctly to macOS equivalents?

I use MS Ergonomic USB keyboard. Also use Karabiner to remap buttons, notably Left Alt/Cmd are swapped. I tried both Left/Right Alt/Cmd, nothing works to change speed.

Is there a Kodi built-in ability to see which buttons are pressed? Note that in all other apps everything works correctly.

**Comment by kambala-decapitator** [comment]
sorry for the false alarm: all works. I simply had to wait for the top+bottom player bars to be gone, only after that changing speed works :)

**Comment by enen92** [comment]
Heads-up: I am considering merging this probably by the end of next week if no objections. So far all feedback has been positive and there isn't a single issue I've seen regarding the switch. I suppose that's good news right? :)

**Comment by enen92** [comment]
Time to say goodbye to our good old sdl :)

**Comment by Memphiz** [comment]
Thx for finishing this - really happy about it :)

**Comment by enen92** [comment]
Great to see you still around :)
Hope life is going well!

**Pull request merged**